File: backend/engine.py
# -*- coding: utf-8 -*-
"""
診断・マッチングエンジン（バックエンドの中核ロジック）。

仕様書 STEP1〜3 を実装:
  STEP1 前提条件 -> 総予算 = 世帯年収 × 7、許容坪単価 = 総予算 / 希望坪数 で絞り込み
  STEP2 価値観診断 -> 20問×5段階回答を5軸(居住性/資産性/施設充実度/立地/広さ)スコアへ集計
  STEP3 タイプ判定 + レーダー用5軸 + マッチ度順 物件TOP3

※スコアの計算式・係数はプロトタイプ用の暫定値。正式なスコアリング定義は
  従DB（後日支給）の確定後に調整します。
"""
import math
import db

TSUBO_PER_SQM = 3.30578  # 1坪 = 約3.30578㎡
AXES = ["liv", "asset", "fac", "loc", "size"]


# ---------------------------------------------------------------- STEP1: 予算
def compute_budget(income_man, area_sqm):
    """income_man: 世帯年収[万円], area_sqm: 希望面積[㎡]"""
    income_man = max(0, float(income_man or 0))
    area_sqm = max(1.0, float(area_sqm or 1))
    max_total = income_man * db.INCOME_MULTIPLIER          # 総予算[万円]
    tsubo = area_sqm / TSUBO_PER_SQM                        # 希望坪数
    max_tsubo_price = max_total / tsubo if tsubo else 0     # 許容坪単価[万円/坪]
    return {
        "incomeMultiplier": db.INCOME_MULTIPLIER,
        "maxTotal": round(max_total),
        "areaSqm": round(area_sqm, 1),
        "tsuboCount": round(tsubo, 2),
        "maxTsuboPrice": round(max_tsubo_price, 1),
    }


# ------------------------------------------------------- STEP2: 5軸スコア集計
def compute_user_axes(answers, questions_doc):
    """answers: {questionId(str/int): value(-2..2)} -> 各軸 0-100 に正規化"""
    answers = answers or {}
    # 軸ごとに raw を集計
    raw = {a: 0.0 for a in AXES}
    count = {a: 0 for a in AXES}
    for q in questions_doc["questions"]:
        a = q["axis"]
        v = answers.get(str(q["id"]), answers.get(q["id"], 0))
        try:
            v = float(v)
        except (TypeError, ValueError):
            v = 0.0
        v = max(-2.0, min(2.0, v))
        if q.get("rev"):
            v = -v
        raw[a] += v
        count[a] += 1
    # 正規化: 各軸 n問 → [-2n, +2n] を 0-100 へ
    user = {}
    for a in AXES:
        n = count[a] or 1
        lo, hi = -2 * n, 2 * n
        user[a] = round((raw[a] - lo) / (hi - lo) * 100) if hi != lo else 50
    return user


# --------------------------------------------------------- STEP3: タイプ判定
def classify_type(user_axes, types):
    """ユーザー5軸ベクトルに最も近い archetype を採用"""
    best, best_dist = None, float("inf")
    for t in types:
        d = math.sqrt(sum((user_axes[a] - t["axis"][a]) ** 2 for a in AXES))
        if d < best_dist:
            best, best_dist = t, d
    return {
        "key": best["key"], "name": best["name"], "emoji": best["emoji"],
        "catch": best["catch"], "desc": best["desc"], "axis": best["axis"],
    }


# --------------------------------------------- STEP3: 物件スコアリング & TOP3
def score_mansions(user_axes, budget, mansions, top_n=3):
    # 重み = ユーザーの軸スコア。全て中立(=50付近)でも比較できるよう微小な下駄を履かせる
    weights = {a: user_axes[a] + 10 for a in AXES}
    wsum = sum(weights.values()) or 1
    max_tsubo = budget["maxTsuboPrice"]

    ranked = []
    affordable_count = 0
    for m in mansions:
        s = m["scores"]
        match_raw = sum(weights[a] * s[a] for a in AXES) / wsum  # 0-100
        affordable = m["tsuboPrice"] <= max_tsubo
        if affordable:
            affordable_count += 1
        ranked.append({
            "id": m["id"], "name": m["name"], "area": m["area"],
            "tsuboPrice": m["tsuboPrice"],                 # 坪単価=直近3ヶ月の成約移動平均(1取引1票)
            "tsuboWindowMonths": m.get("tsuboWindowMonths"),  # 算定に使った窓(基本3/拡大時はその月数)
            "txInWindow": m.get("txInWindow"),             # 窓内の成約件数
            "trendPct": m.get("trendPct"),                 # 値上がり率(資産性の裏付け)
            "txCount": m.get("txCount"),                   # 累計成約件数
            "medianSqm": m.get("medianSqm"),               # 成約面積の中央値
            "repLayout": m.get("repLayout", ""),           # 代表間取り
            # 従DB(スペック)由来
            "station": m.get("station"), "walkMin": m.get("walkMin"),
            "builtYear": m.get("builtYear"), "ageYears": m.get("ageYears"),
            "seismic": m.get("seismic"), "floors": m.get("floors"),
            "totalUnits": m.get("totalUnits"),
            "photoUrl": m.get("photoUrl"),
            "facilities": m.get("facilities", {}),
            "specsPending": m.get("specsPending", False),
            "scores": s,
            "estTotal": round(m["tsuboPrice"] * budget["tsuboCount"]),  # 希望面積での概算総額[万円]
            "matchPct": max(1, min(99, round(match_raw))),
            "overBudget": not affordable,
        })

    # 予算内を優先し、その中でマッチ度降順。予算内が3件未満なら予算オーバーで補完。
    ranked.sort(key=lambda x: (x["overBudget"], -x["matchPct"]))
    return ranked[:top_n], affordable_count


# ------------------------------------------------------------------- まとめ
def diagnose(payload):
    age = payload.get("age")
    income = payload.get("income")
    area_sqm = payload.get("areaSqm")
    answers = payload.get("answers", {})

    questions_doc = db.get_questions_doc()
    types = db.get_types()
    mansions = db.get_mansions()

    budget = compute_budget(income, area_sqm)
    user_axes = compute_user_axes(answers, questions_doc)
    type_ = classify_type(user_axes, types)
    matches, affordable_count = score_mansions(user_axes, budget, mansions)

    return {
        "input": {"age": age, "income": income, "areaSqm": area_sqm},
        "budget": budget,
        "userAxis": user_axes,
        "axesMeta": questions_doc["axes"],
        "type": type_,
        "matches": matches,
        "affordableCount": affordable_count,
        "totalCount": len(mansions),
        "lineUrl": db.LINE_URL,
        "isMock": True,
    }

File: docs/engine.js
/* ===== 湾岸マッチング 診断エンジン v2（ブラウザ版・依存ゼロ） =====
   - 6軸(居住性/資産性/施設/立地/広さ/眺望)の価値観診断
   - 提案は「現在募集中の住戸(listings)」を、定量条件で絞り込み→6軸マッチ度でTOP3
   - 予算は直接入力（旧: 世帯年収×7 は廃止）
   ※建物5軸スコアは mansions.json、眺望は住戸の階数から算出済(listings.viewScore)。 */
(function (global) {
  "use strict";
  var TSUBO_PER_SQM = 3.30578;
  var AXES = ["liv", "asset", "fac", "loc", "size", "view"];

  // STEP2: 24問 → 6軸 0-100
  function computeUserAxes(answers, questions) {
    answers = answers || {};
    var raw = {}, count = {};
    AXES.forEach(function (a) { raw[a] = 0; count[a] = 0; });
    questions.forEach(function (q) {
      var a = q.axis;
      if (raw[a] === undefined) return;
      var v = answers[q.id];
      if (v === undefined) v = answers[String(q.id)];
      v = Number(v) || 0;
      v = Math.max(-2, Math.min(2, v));
      if (q.rev) v = -v;
      raw[a] += v;
      count[a] += 1;
    });
    var user = {};
    AXES.forEach(function (a) {
      var n = count[a] || 1, lo = -2 * n, hi = 2 * n;
      user[a] = hi !== lo ? Math.round((raw[a] - lo) / (hi - lo) * 100) : 50;
    });
    return user;
  }

  // STEP3: タイプ判定（最近傍 archetype・6軸ユークリッド距離）
  function classifyType(userAxes, types) {
    var best = null, bestDist = Infinity;
    types.forEach(function (t) {
      var d = 0;
      AXES.forEach(function (a) { var diff = userAxes[a] - (t.axis[a] || 0); d += diff * diff; });
      d = Math.sqrt(d);
      if (d < bestDist) { bestDist = d; best = t; }
    });
    return { key: best.key, name: best.name, emoji: best.emoji, catch: best.catch, desc: best.desc, axis: best.axis };
  }

  // 間取りをカテゴリ化（フォームのチェックボックスと突合：1R/1LDK/2LDK/3LDK/4LDK+）
  function layoutCat(s) {
    if (!s) return null;
    if (/(ワンルーム|^1R|^1K|^1DK|^1SK|^STUDIO)/i.test(s)) return "1R";
    var m = s.match(/^([1-9])/);
    if (!m) return "other";
    var n = +m[1];
    if (n >= 4) return "4LDK+";
    return n + "LDK";
  }

  // 住戸の6軸ベクトル（建物5軸 + 住戸の眺望）
  function listingVector(L, building) {
    var s = (building && building.scores) || {};
    return {
      liv: s.liv != null ? s.liv : 50,
      asset: s.asset != null ? s.asset : 50,
      fac: s.fac != null ? s.fac : 50,
      loc: s.loc != null ? s.loc : 50,
      size: s.size != null ? s.size : 50,
      view: L.viewScore != null ? L.viewScore : 50,
    };
  }

  // STEP3: 現在募集中の住戸を 定量条件で絞り込み → 6軸マッチ度でTOP3
  //   conds = { budget(万,任意), areaMin(㎡,任意), layouts([],任意), ageMax(年,任意), walkMax(分,任意) }
  function scoreListings(userAxes, conds, listings, mansionById, topN) {
    topN = topN || 3;
    conds = conds || {};
    var weights = {}, wsum = 0;
    AXES.forEach(function (a) { weights[a] = userAxes[a] + 10; wsum += weights[a]; });
    if (!wsum) wsum = 1;
    var hasLayout = conds.layouts && conds.layouts.length > 0;

    var pass = [];
    listings.forEach(function (L) {
      var b = mansionById[L.bid] || {};
      // ---- ハード条件（満たさなければ除外）----
      if (conds.budget && L.price > conds.budget) return;
      if (conds.areaMin && (L.sqm == null || L.sqm < conds.areaMin)) return;
      if (hasLayout && (!L.layout || conds.layouts.indexOf(layoutCat(L.layout)) < 0)) return;
      if (conds.ageMax != null && b.ageYears != null && b.ageYears > conds.ageMax) return;
      if (conds.walkMax != null && b.walkMin != null && b.walkMin > conds.walkMax) return;
      // ---- ソフト：6軸マッチ度 ----
      var vec = listingVector(L, b);
      var matchRaw = 0;
      AXES.forEach(function (a) { matchRaw += weights[a] * vec[a]; });
      matchRaw = matchRaw / wsum;
      pass.push({
        listing: L, building: b, vec: vec,
        matchPct: Math.max(1, Math.min(99, Math.round(matchRaw))),
      });
    });
    pass.sort(function (a, b) {
      if (b.matchPct !== a.matchPct) return b.matchPct - a.matchPct;
      return (b.listing.viewScore || 0) - (a.listing.viewScore || 0);
    });
    // 同一建物の重複を除き、別々の建物でTOP3（選択肢のバリエーション確保）
    var seen = {}, uniq = [];
    pass.forEach(function (p) {
      var k = p.listing.bid || p.listing.bkey;
      if (seen[k]) return;
      seen[k] = 1; uniq.push(p);
    });
    return { matches: uniq.slice(0, topN), candidateCount: pass.length, buildingCount: uniq.length };
  }

  // まとめ
  function diagnose(input, data, opts) {
    opts = opts || {};
    var userAxes = computeUserAxes(input.answers, data.questions);
    var type = classifyType(userAxes, data.types);
    var mById = {};
    (data.mansions || []).forEach(function (m) { mById[m.id] = m; });
    var conds = input.conds || {};
    var r = scoreListings(userAxes, conds, data.listings || [], mById, 3);
    return {
      conds: conds,
      userAxis: userAxes,
      axesMeta: data.axes,
      type: type,
      matches: r.matches,
      candidateCount: r.candidateCount,
      totalListings: (data.listings || []).length,
      lineUrl: opts.lineUrl || "",
    };
  }

  global.WANGAN = { diagnose: diagnose, scoreListings: scoreListings, computeUserAxes: computeUserAxes, classifyType: classifyType, layoutCat: layoutCat, AXES: AXES, TSUBO_PER_SQM: TSUBO_PER_SQM };
})(window);
